lib/ansible/plugins/action/__init__.py

```diff
@@ -1277,43 +1278,39 @@ def _low_level_execute_command(self, cmd, sudoable=True, in_data=None, executabl
-        #     display.debug("_low_level_execute_command(): no command, exiting")
-        #     return dict(stdout='', stderr='', rc=254)
-
-        if chdir:
-            display.debug("_low_level_execute_command(): changing cwd to %s for this command" % chdir)
-            cmd = self._connection._shell.append_command('cd %s' % chdir, cmd)
 
         # https://github.com/ansible/ansible/issues/68054
         if executable:
             self._connection._shell.executable = executable
 
         ruser = self._get_remote_user()
         buser = self.get_become_option('become_user')
-        if (sudoable and self._connection.become and  # if sudoable and have become
-                resource_from_fqcr(self._connection.transport) != 'network_cli' and  # if not using network_cli
-                (C.BECOME_ALLOW_SAME_USER or (buser != ruser or not any((ruser, buser))))):  # if we allow same user PE or users are different and either is set
-            display.debug("_low_level_execute_command(): using become for this command")
-            cmd = self._connection.become.build_become_command(cmd, self._connection._shell)
-
-        if self._connection.allow_executable:
-            if executable is None:
-                executable = self._play_context.executable
-                # mitigation for SSH race which can drop stdout (https://github.com/ansible/ansible/issues/13876)
-                # only applied for the default executable to avoid interfering with the raw action
-                cmd = self._connection._shell.append_command(cmd, 'sleep 0')
-            if executable:
-                cmd = executable + ' -c ' + shlex.quote(cmd)
-
-        display.debug("_low_level_execute_command(): executing: %s" % (cmd,))
-
-        # Change directory to basedir of task for command execution when connection is local
-        if self._connection.transport == 'local':
-            self._connection.cwd = to_bytes(self._loader.get_basedir(), errors='surrogate_or_strict')
-
-        rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
+        use_become = bool(
+            sudoable and
+            # FUTURE: this should be moved into the network_cli prepare_command
+            # function once 2.16 is the min version for that collection.
+            resource_from_fqcr(self._connection.transport) != 'network_cli' and
+            (
+                C.BECOME_ALLOW_SAME_USER or
+                (buser != ruser or not any((ruser, buser)))
+            )
+        )
+
+        # Allow the connection plugin to prepare the final command and input
+        # data in case it ahs any custom rules or special handling
+        cmd_info = PrepareCommandInfo(
+            cmd=cmd,
+            in_data=in_data,
+            use_become=use_become,
+            executable=executable,
+            chdir=chdir,
+            ansible_loader_basedir=self._loader.get_basedir(),
```

Comment:
I was hoping to avoid adding this attribute but a connection Cannot use `self._play_context.get_loader().get_basedir()` as the play context does have the `get_loader()` method but it is no longer set.